src/main/kotlin/com/mdrobnak/lalrpop/intentions/AddExplicitTypeIntention.kt

```diff
@@ -0,0 +1,64 @@
+package com.mdrobnak.lalrpop.intentions
+
+import com.intellij.codeInsight.intention.IntentionAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiFile
+import com.mdrobnak.lalrpop.LpLanguage
+import com.mdrobnak.lalrpop.psi.LpAlternative
+import com.mdrobnak.lalrpop.psi.LpElementFactory
+import com.mdrobnak.lalrpop.psi.LpNonterminal
+import com.mdrobnak.lalrpop.psi.getContextAndResolveType
+
+class AddExplicitTypeIntention : IntentionAction {
+    override fun startInWriteAction(): Boolean = true
+
+    override fun getText(): String = "Add explicit type"
+
+    override fun getFamilyName(): String = "Add explicit type"
+
+    override fun isAvailable(project: Project, editor: Editor?, file: PsiFile?): Boolean {
+        if (editor == null || file == null || file.language != LpLanguage) return false
+
+        var element = file.findElementAt(editor.caretModel.primaryCaret.offset)
+        while (element != null) {
+            if (element is LpNonterminal) return true
+            element = element.parent
+        }
+        return false
+    }
+
+    override fun invoke(project: Project, editor: Editor?, file: PsiFile?) {
+        if (editor == null || file == null || file.language != LpLanguage) return
+
+        var element = file.findElementAt(editor.caretModel.primaryCaret.offset)
+
+        var nonterminal: LpNonterminal? = null
+        var alternativeToUse: LpAlternative? = null
+
+        while (element != null) {
+            if (element is LpAlternative) {
+                if (element.action == null) // TODO: maybe get from the rust plugin?
```

Comment:
This should definitely get resolved; the intention currently doesn't really work for nonterminals with action code.

Reply:
Am so close but generics gave me a headache

Reply:
What's the type resolution bug you found?

Reply:
Happened after I removed `private` from https://github.com/Mcat12/intellij-lalrpop/pull/26/files#diff-5277a2f0d1a63609b83c3e82765f722756d0a354aebd19ff4925d5b98523f858L191 and didn't read through https://github.com/Mcat12/intellij-lalrpop/pull/26/files#diff-2905bcab179a961af5375f1158d30c3e950103907a7693247ccb47541bf8a565L15-L17 properly, and it ended up just using the text even for type refs.